sql: rollback transaction eagerly upon error

rafiss opened the pull request:
### sql: rollback transaction eagerly upon error

If we have not used any savepoints, it's safe to rollback the
transaction immediately when a non-retryable error is encountered.

This releases locks sooner, which is useful so we avoid needing to wait
until the client sends a ROLLBACK statement.

Release note (bug fix): Transactions that enter the aborted state will
now release locks they are holding immediately, as long as there is no
SAVEPOINT active in the transaction.

### sql: make transaction_timeout abort transaction eagerly

Previously, the transaction_timeout setting would only cause the
transaction to abort if the timeout had already expired by the time the
next statement is received by CRDB. Now we create a timer that fires if
the timeout expires while we are waiting for the next statement.

There's no release note, since up until the previous commit, which
changed the behavior so transactions eagerly release locks, there was no
functional difference caused by lazily timing out the transaction.

Release note: None

###  sql: add test for implicit txn holding locks across retries

Release note: None

---

fixes https://github.com/cockroachdb/cockroach/issues/140234

rafiss committed:
sql: rollback transaction eagerly upon error

If we have not used any savepoints, it's safe to rollback the
transaction immediately when a non-retryable error is encountered.

This releases locks sooner, which is useful so we avoid needing to wait
until the client sends a ROLLBACK statement.

Release note (bug fix): Transactions that enter the aborted state will
now release locks they are holding immediately, as long as there is no
SAVEPOINT active in the transaction.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/140160)
<!-- Reviewable:end -->

michae2 reviewed (commented):
Looks pretty good!

Some test ideas:
- something with cockroach_restart
- something with pausable portals
- something with implicit txn retry

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-:-OHxlaMa1M7fqqR0vMg9:b-86yvag)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rafiss)

___
*[`pkg/sql/conn_executor_savepoints.go` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-OHxkzxmAJMlbzKm5pkl:-OHxkzxmAJMlbzKm5pkm:b3o80nd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a34004f2ea6a1d4d5bd280fe006a42ff102a1e09/pkg/sql/conn_executor_savepoints.go#L34)):*
> ```Go
> 	ctx context.Context, s *tree.Savepoint, res RestrictedCommandResult,
> ) (fsm.Event, fsm.EventPayload, error) {
> 	ex.state.mu.hasSavepoints = true
> ```

Would we want to set `hasSavepoints` to false somewhere in execRelease to allow the eager rollback of a transaction that fails after all created savepoints have been released?


<!-- Sent from Reviewable.io -->

rafiss committed:
sql: make transaction_timeout abort transaction eagerly

Previously, the transaction_timeout setting would only cause the
transaction to abort if the timeout had already expired by the time the
next statement is received by CRDB. Now we create a timer that fires if
the timeout expires while we are waiting for the next statement.

There's no release note, since up until the previous commit, which
changed the behavior so transactions eagerly release locks, there was no
functional difference caused by lazily timing out the transaction.

Release note: None

rafiss committed:
sql: add test for implicit txn holding locks across retries

Release note: None

rafiss reviewed (commented):
>    something with cockroach\_restart

I added a test with savepoints. I don't think cockroach_restart needs any other special testing, but I could be missing something, so just let me know what else you had in mind.

>  something with pausable portals

I was thinking this would be covered by our metamorphic testing. But I can add a test for a specific scenario if you can think of one.

>   something with implicit txn retry

Added this test.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-:-OIlAExj1YKuV1dsYC64:b-wvpsna)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/conn_executor_savepoints.go` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-OHxkzxmAJMlbzKm5pkl:-OIlAc_w-LcxtUGnl_y7:b-ij0a48) ([raw file](https://github.com/cockroachdb/cockroach/blob/a34004f2ea6a1d4d5bd280fe006a42ff102a1e09/pkg/sql/conn_executor_savepoints.go#L34)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Would we want to set `hasSavepoints` to false somewhere in execRelease to allow the eager rollback of a transaction that fails after all created savepoints have been released?
</blockquote></details>

done. good point.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140160)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/conn_fsm.go` line 418 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-OIrM2Yl2y6PZ76fW7te:-OIrM2Yl2y6PZ76fW7tf:blsdype) ([raw file](https://github.com/cockroachdb/cockroach/blob/b898715725250ce5f9aefc6fd4997f5768e7d6b6/pkg/sql/conn_fsm.go#L418)):*
> ```Go
> 					ts.mu.Lock()
> 					defer ts.mu.Unlock()
> 					if !ts.mu.hasSavepoints {
> ```

Something occurred to me: if this is a nested connExecutor created by `InternalExecutor.initConnEx`, sharing a transaction with its parent connExecutor, then will it know about the savepoints in its parent connExecutor?

Maybe we also need to check that this is a top-level connExecutor? Or we need to add savepoint info to `InternalExecutor.extraTxnState`?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140160)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/conn_fsm.go` line 418 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-OIrM2Yl2y6PZ76fW7te:-OIsrGPw4L-QTa_lQbYG:b-dkqw2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/b898715725250ce5f9aefc6fd4997f5768e7d6b6/pkg/sql/conn_fsm.go#L418)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Something occurred to me: if this is a nested connExecutor created by `InternalExecutor.initConnEx`, sharing a transaction with its parent connExecutor, then will it know about the savepoints in its parent connExecutor?

Maybe we also need to check that this is a top-level connExecutor? Or we need to add savepoint info to `InternalExecutor.extraTxnState`?
</blockquote></details>

When we're running an internal executor delegated from a parent transaction, the connExecutor uses a different state machine. It's further down below: https://github.com/cockroachdb/cockroach/blob/5a8f03f9c410b3951a56078c933d576cf9694f0e/pkg/sql/conn_fsm.go#L704

The transaction is never rolled back in those state transitions.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> Reviewed 7 of 7 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140160)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/conn_fsm.go` line 418 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-OIrM2Yl2y6PZ76fW7te:-OJ-RgPd9UXpDU3svIas:bszelp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b898715725250ce5f9aefc6fd4997f5768e7d6b6/pkg/sql/conn_fsm.go#L418)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

When we're running an internal executor delegated from a parent transaction, the connExecutor uses a different state machine. It's further down below: https://github.com/cockroachdb/cockroach/blob/5a8f03f9c410b3951a56078c933d576cf9694f0e/pkg/sql/conn_fsm.go#L704

The transaction is never rolled back in those state transitions.
</blockquote></details>

Great, thank you!
___
*[`pkg/sql/conn_executor_test.go` line 1905 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-OJ-bhEECOTXyA0bVyoR:-OJ-bhEECOTXyA0bVyoS:b-qttsw8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b898715725250ce5f9aefc6fd4997f5768e7d6b6/pkg/sql/conn_executor_test.go#L1905)):*
> ```Go
> 		err = conn1.QueryRowContext(ctx, `SELECT v FROM t WHERE k = 4`).Scan(&v)
> 		require.NoError(t, err)
> 		require.Equal(t, 400, v)
> ```

one idea for a cockroach_restart test is something like:

```sql
-- conn 1
INSERT INTO t VALUES (5, 5), (6, 6);
BEGIN;
SAVEPOINT cockroach_restart;
SELECT * FROM t WHERE k = 5;
    -- conn 2 (to break serializability)
    UPDATE t SET v = 50 WHERE k = 5;
    SELECT * FROM t WHERE k = 6;
-- conn 1
UPDATE t SET v = 60 WHERE k = 6;
    -- conn 2 (blocks, to prove locks are still held)
    UPDATE t SET v = v + 600 WHERE k = 6;
-- conn 1 (should fail with RETRY_SERIALIZABLE)
RELEASE SAVEPOINT cockroach_restart;
-- simulate the advanced retry
ROLLBACK TO SAVEPOINT cockroach_restart;
SAVEPOINT cockroach_restart;
SELECT * FROM t WHERE k = 5;
UPDATE t SET v = 61 WHERE k = 6;
RELEASE SAVEPOINT cockroach_restart;
    -- conn 2 (to prove that ordering was correct and locks are now released)
    SELECT * FROM t WHERE k = 6;
-- conn 1
COMMIT;
```
___
*[`pkg/sql/conn_executor_test.go` line 1916 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-OJ-ul-tCAvKHcCUZAB0:-OJ-ul-tCAvKHcCUZAB1:bscux7x) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d4bcf258af9de01595ba9f82f41561744f36d4b/pkg/sql/conn_executor_test.go#L1916)):*
> ```Go
> // layer, but for now this is the expected behavior.
> // See https://github.com/cockroachdb/cockroach/issues/117020.
> func TestRetriableErrorDuringTransactionHoldsLocks(t *testing.T) {
> ```

Thanks for adding this!
___
*[`pkg/sql/conn_executor_exec.go` line 186 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-OJ-str6-NfR1tCV3HMT:-OJ-str6-NfR1tCV3HMU:b-t9fn2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/7551e610c4f89838d789db4b4608f98beecc3db0/pkg/sql/conn_executor_exec.go#L186)):*
> ```Go
> 	ex.startIdleInSessionTimeout()
> 
> 	if ex.sessionData().IdleInTransactionSessionTimeout > 0 {
> ```

If both IdleInTransactionSessionTimeout and txnTimeoutRemaining are > 0, do we need to set both timers? Seems like we only need to set the one that will fire first?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140160)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/conn_executor_exec.go` line 186 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-OJ-str6-NfR1tCV3HMT:-OJU6aa49AOowSXK7gOn:bvdvboo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7551e610c4f89838d789db4b4608f98beecc3db0/pkg/sql/conn_executor_exec.go#L186)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

If both IdleInTransactionSessionTimeout and txnTimeoutRemaining are > 0, do we need to set both timers? Seems like we only need to set the one that will fire first?
</blockquote></details>

the two timers have slightly different behaviors. as per the commend i added below:
```
				// The transaction_timeout setting should move the transaction to the
				// aborted state.
				// NOTE: In Postgres, the transaction_timeout causes the entire session
				// to be terminated. We intentionally diverge from that behavior.
```

however, the IdleInTransactionSessionTimeout timer causes the session to be terminated.
___
*[`pkg/sql/conn_executor_test.go` line 1905 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-OJ-bhEECOTXyA0bVyoR:-OJU6s1q64i6zgk1qlge:b-vs7g9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b898715725250ce5f9aefc6fd4997f5768e7d6b6/pkg/sql/conn_executor_test.go#L1905)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

one idea for a cockroach_restart test is something like:

```sql
-- conn 1
INSERT INTO t VALUES (5, 5), (6, 6);
BEGIN;
SAVEPOINT cockroach_restart;
SELECT * FROM t WHERE k = 5;
    -- conn 2 (to break serializability)
    UPDATE t SET v = 50 WHERE k = 5;
    SELECT * FROM t WHERE k = 6;
-- conn 1
UPDATE t SET v = 60 WHERE k = 6;
    -- conn 2 (blocks, to prove locks are still held)
    UPDATE t SET v = v + 600 WHERE k = 6;
-- conn 1 (should fail with RETRY_SERIALIZABLE)
RELEASE SAVEPOINT cockroach_restart;
-- simulate the advanced retry
ROLLBACK TO SAVEPOINT cockroach_restart;
SAVEPOINT cockroach_restart;
SELECT * FROM t WHERE k = 5;
UPDATE t SET v = 61 WHERE k = 6;
RELEASE SAVEPOINT cockroach_restart;
    -- conn 2 (to prove that ordering was correct and locks are now released)
    SELECT * FROM t WHERE k = 6;
-- conn 1
COMMIT;
```
</blockquote></details>

nice! let me give this a try


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
thanks for the reviews! this is ready for another look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-:-OJV3UxT8P5bHwQEuiKE:b-3k29w8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @michae2)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Just one suggestion. This is great!

> Reviewed 4 of 4 files at r5, 3 of 3 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-:-OJszzQ47ab-aQZcT0Mq:b-k7ocub)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/conn_executor_exec.go` line 186 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-OJ-str6-NfR1tCV3HMT:-OJt258PEDuphCrSCl8y:bmmqevl) ([raw file](https://github.com/cockroachdb/cockroach/blob/7551e610c4f89838d789db4b4608f98beecc3db0/pkg/sql/conn_executor_exec.go#L186)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

the two timers have slightly different behaviors. as per the commend i added below:
```
				// The transaction_timeout setting should move the transaction to the
				// aborted state.
				// NOTE: In Postgres, the transaction_timeout causes the entire session
				// to be terminated. We intentionally diverge from that behavior.
```

however, the IdleInTransactionSessionTimeout timer causes the session to be terminated.
</blockquote></details>

Ah, I missed that. Thanks!
___
*[`pkg/sql/conn_executor_test.go` line 1936 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-OJt-wkwC8e8u2gMmcSd:-OJt-wkwC8e8u2gMmcSe:bkt87qi) ([raw file](https://github.com/cockroachdb/cockroach/blob/c834ac7349f4e24001930a5099f1e8b4fea822d7/pkg/sql/conn_executor_test.go#L1936)):*
> ```Go
> 
> 		_, err = conn1.ExecContext(ctx, `RELEASE SAVEPOINT cockroach_restart`)
> 		require.ErrorContains(t, err, "failed preemptive refresh due to encountered recently written committed value")
> ```

nit: Because we're not letting the conn2 update block the entire time, I think it would be a slightly better test to try the conn2 update right after releasing the cockroach_restart savepoint. This would validate that locks are still held after releasing the cockroach_restart savepoint.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
tftr!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-:-OJz7n6n0gKmadfWAXUc:bjpalj3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @michae2)

___
*[`pkg/sql/conn_executor_test.go` line 1936 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-OJt-wkwC8e8u2gMmcSd:-OJz7dEn7rDj9I1nmAJ5:btpapyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c834ac7349f4e24001930a5099f1e8b4fea822d7/pkg/sql/conn_executor_test.go#L1936)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: Because we're not letting the conn2 update block the entire time, I think it would be a slightly better test to try the conn2 update right after releasing the cockroach_restart savepoint. This would validate that locks are still held after releasing the cockroach_restart savepoint.
</blockquote></details>

good point! added the assertion after too


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks!!!

> Reviewed 3 of 7 files at r2, 4 of 4 files at r8, 3 of 3 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140160#-:-OK6x-k27_29FGodCP1z:b-n7iw47)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->
